BiometricLocker/BiometricLocker.swift

```diff
@@ -70,49 +99,95 @@ open class BiometricLocker {
         return false
     }
 
-    /// Locks the app.
-    ///
-    /// - Parameter date: Tells the locker when the app was sent to the background (or any other situation, when applicable), so that we can calculate if the app should be locked.
-    open func lock(at date: Date = Date()) {
+    /// Store the NotificationCenter observers for deallocation.
+    private var notificationObservers = [NSObjectProtocol]()
+
+    public init(localizedReason: String, automaticallyLocksOnBackgroundOrQuit: Bool = true, withUnlockedTimeAllowance: Bool = true) {
+        self.localizedReason = localizedReason
+        
+        if !withUnlockedTimeAllowance {
+            self.unlockedTimeAllowance = 0
+        }
+
+        if automaticallyLocksOnBackgroundOrQuit {
+            self.notificationObservers.append(NotificationCenter.default.addObserver(forName: .UIApplicationDidEnterBackground, object: nil, queue: .main) { _ in
+                // If app goes into background, we start the clocks to lock the app.
+                self.lock()
+            })
+
+            self.notificationObservers.append(NotificationCenter.default.addObserver(forName: .UIApplicationWillTerminate, object: nil, queue: .main) { _ in
+                // If the app is killed, lock it instantly.
+                self.lock(.now)
+            })
+        }
+    }
+
+    deinit {
+        self.notificationObservers.forEach { observer in
+            NotificationCenter.default.removeObserver(observer)
+        }
+    }
+
+    /**
+     Tells the biometric locker to lock the app. Defaults to the currently defined time allowance (5min by default).
+
+     - Important: To change when the app locks after having called this, simply call `lock` again with a different allowance, or tell the app to `lock(.afterTimeInterval(timeInterval)`, as it supercedes the current unlocked time allowance. `unlockedTimeAllowance` is only checked when we ask the BiometricLocker whether the app is locked or not. Changing it before or after calling `lock` should work just fine.
+     - Important:  Keep in mind that changing `unlockedTimeAllowance` after setting a custom time interval will not work as expected.
+     - Parameter when: An enum defining when the app should lock. Now, after the time allowance, or after a custom time interval. Keep in mind
+     */
+    public func lock(_ when: LockingTime = .afterTimeAllowance) {
         // if we are already deactivated, return. Otherwise you can just kill the app and try again to bypass touch ID.
         if self.isLocked { return }
+
+        let date: Date
+        switch when {
+        case .now:
+            date = Date.distantPast
+        case .afterTimeAllowance:
+            date = Date()
+        case .afterTimeInterval(let interval):
+            date = Date(timeIntervalSinceNow: (interval - self.unlockedTimeAllowance))
```

Comment:
still wondering about unlocked time allowance being greater than interval

Reply:
Best way to be sure: tests! 🎉 Adding some.